### PR TITLE
[TDF] Simplify action creation logic

### PR DESCRIPTION
**Pull request by bluehood**
- removed nested template struct SimpleAction
- TDataFrameInterface::SimpleAction::BuildAndBook is now TDataFrameInterface::BuildAndBook
- BuildAndBook now only takes one template parameter instead of three
- the default type of histogram weights is now `void` (which signals that we are
  not filling a weighted histogram)
- `enum EActionType` has been replaced by `namespace ActionTypes`

With a few exceptions, actions are now created by calling `CreateAction`, which
in turn calls `BuildAndBook`. Two template overloads of `CreateAction` are used
to decide whether to perform runtime type guessing or not. Multiple overloads of
`BuildAndBook` take care of building the correct operation for each action.

### CODE REVIEW

File: tree/treeplayer/inc/ROOT/TDataFrame.hxx [hunk 1]
```diff
@@ -278,7 +278,12 @@ public:
    }
 };
 
-enum class EActionType : short { kHisto1D, kMin, kMax, kMean };
+namespace ActionTypes {
+   struct Histo1D_t {};
```
Comment (pcanal) [inline tree/treeplayer/inc/ROOT/TDataFrame.hxx]: Please use 'just' 'Histo1D'.  The _t suffix is reserved for typedefs.  thanks.

Comment (bluehood) [inline tree/treeplayer/inc/ROOT/TDataFrame.hxx]: done

File: tree/treeplayer/inc/ROOT/TDataFrame.hxx [hunk 2]
```diff
@@ -709,169 +717,131 @@ private:
...
+         // else if (typeCode == 'b') { return BuildAndBook<UChar_t>(bl, r, nSlots); }
+         // else if (typeCode == 'S') { return BuildAndBook<Short_t>(bl, r, nSlots, (AT_t*)nullptr); }
+         // else if (typeCode == 's') { return BuildAndBook<UShort_t>(bl, r, nSlots, (AT_t*)nullptr); }
+         // else if (typeCode == 'i') { return BuildAndBook<int>(bl, r, nSlots, (AT_t*)nullptr); }
+         // else if (typeCode == 'F') { return BuildAndBook<float>(bl, r, nSlots, (AT_t*)nullptr); }
+         // else if (typeCode == 'L') { return BuildAndBook<Long64_t>(bl, r, nSlots, (AT_t*)nullptr); }
+         // else if (typeCode == 'l') { return BuildAndBook<ULong64_t>(bl, r, nSlots, (AT_t*)nullptr); }
+         // else if (typeCode == 'O') { return BuildAndBook<bool>(bl, r, nSlots, (AT_t*)nullptr); }
       } else {
          std::string typeName = branchEl->GetTypeName();
```
Comment (pcanal) [inline tree/treeplayer/inc/ROOT/TDataFrame.hxx]: Please consider comparing the TClass returned by TBranch::GetExpectedType.  (and use a static TClassRef to cache the 'target' TClass.

File: tree/treeplayer/inc/ROOT/TDataFrame.hxx [hunk 3]
```diff
@@ -709,169 +717,131 @@ private:
...
+         else if (type_id == typeid(int)) { return BuildAndBook<int>(bl, r, nSlots, (AT_t*)nullptr); }
+         else if (type_id == typeid(double)) { return BuildAndBook<double>(bl, r, nSlots, (AT_t*)nullptr); }
+         else if (type_id == typeid(std::vector<double>)) { return BuildAndBook<std::vector<double>>(bl, r, nSlots, (AT_t*)nullptr); }
+         else if (type_id == typeid(std::vector<float>)) { return BuildAndBook<std::vector<float>>(bl, r, nSlots, (AT_t*)nullptr); }
       }
       // real branch
       auto branchEl = dynamic_cast<TBranchElement *>(branch);
       if (!branchEl) { // This is a fundamental type
-         auto title    = branch->GetTitle();
+         auto title = branch->GetTitle();
```
Comment (pcanal) [inline tree/treeplayer/inc/ROOT/TDataFrame.hxx]: Please consider comparing the TDataType rned by TBranch::GetExpectedType. (and use a static TClassRef to cache the 'target' TClass.

Comment (dpiparo) [inline tree/treeplayer/inc/ROOT/TDataFrame.hxx]: This makes sense indeed: thanks.

Comment (dpiparo) [inline tree/treeplayer/inc/ROOT/TDataFrame.hxx]: I propose to refine this aspect of the type guessing logic in another PR as the current one is not related to this particular aspect.

**Comment by dpiparo** [comment]
Merged. The comments of Philippe about using the type system rather than the names of the types will be addressed in a separate PR.

**Pull request closed without merging**